plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClient.java

```diff
@@ -613,18 +677,53 @@ public void setNull(PreparedStatement statement, int index)
         };
     }
 
-    private static LongReadFunction oracleTimestampReadFunction()
+    private static String getOracleBindExpression(int precision)
+    {
+        if (precision == 0) {
+            return "TO_TIMESTAMP(?, 'SYYYY-MM-DD HH24:MI:SS')";
+        }
+        if (precision <= 2) {
+            return "TO_TIMESTAMP(?, 'SYYYY-MM-DD HH24:MI:SS.FF')";
+        }
+        else {
+            return format("TO_TIMESTAMP(?, 'SYYYY-MM-DD HH24:MI:SS.FF%d')", precision);
+        }
+    }
+
+    private static LongReadFunction oracleTimestampReadFunction(TimestampType timestampType)
     {
         return (resultSet, columnIndex) -> {
             LocalDateTime timestamp = resultSet.getObject(columnIndex, LocalDateTime.class);
             // Adjust years when the value is B.C. dates because Oracle returns +1 year unless converting to string in their server side
             if (timestamp.getYear() <= 0) {
                 timestamp = timestamp.minusYears(1);
             }
-            return timestamp.toInstant(ZoneOffset.UTC).toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
+            return toTrinoTimestamp(timestampType, timestamp);
         };
     }
 
+    private static ObjectReadFunction oracleLongTimestampReadFunction(TimestampType timestampType)
+    {
+        verifyLongTimestampPrecision(timestampType.getPrecision());
+        return ObjectReadFunction.of(
+                LongTimestamp.class,
+                (resultSet, columnIndex) ->
+                {
+                    LocalDateTime timestamp = resultSet.getObject(columnIndex, LocalDateTime.class);
+                    // Adjust years when the value is B.C. dates because Oracle returns +1 year unless converting to string in their server side
+                    if (timestamp.getYear() <= 0) {
+                        timestamp = timestamp.minusYears(1);
+                    }
+                    return toLongTrinoTimestamp(timestampType, timestamp);
+                });
+    }
+
+    private static void verifyLongTimestampPrecision(int precision)
+    {
+        checkArgument(precision > MAX_SHORT_PRECISION && precision <= MAX_ORACLE_TIMESTAMP_PRECISION,
+                "Precision is out of range: %s", precision);
+    }
```

Comment:
So for queries which were running previously would it fail now ?

Reply:
I think no, this check will be applied only after rounding:
int precision = min(timestampType.getPrecision(), MAX_ORACLE_TIMESTAMP_PRECISION);